File: packs/ai/app/components/folio/ai/console/text_suggestions_group/controls_component.rb
# frozen_string_literal: true

class Folio::Ai::Console::TextSuggestionsGroup::ControlsComponent < Folio::Console::ApplicationComponent
  def initialize
  end

  private
    def component_data
      stimulus_controller("f-ai-c-text-suggestions-group-controls",
                          action: {
                            "f-ai-c-text-suggestions-group:state": "onGroupState",
                          })
    end

    def button_data
      stimulus_data(action: { click: "generate" },
                    target: "button")
    end

    def close_data
      stimulus_data(action: { click: "close" },
                    target: "close")
    end

    def generate_all_label
      label(:generate_all_label)
    end

    def close_label
      label(:close_all_label)
    end

    def label(key)
      I18n.t(key, scope: "folio.ai.console.text_suggestions_group_component")
    end
end

File: packs/ai/app/components/folio/ai/console/text_suggestions_group/controls_component.js
(() => {
  const STATE_IDLE = 'idle'
  const STATE_LOADING = 'loading'

  const registerControlsController = () => {
    window.Folio.Stimulus.register('f-ai-c-text-suggestions-group-controls', class extends window.Stimulus.Controller {
      static targets = ['button', 'close']

      connect () {
        this.groupState = this.initialGroupState
        this.syncControls()
      }

      generate (event) {
        this.handleActionEvent(event)
        if (this.isLoading) return

        this.dispatch('generate', { bubbles: true })
      }

      close (event) {
        this.handleActionEvent(event)

        this.dispatch('close', { bubbles: true })
      }

      onGroupState (event) {
        this.groupState = event?.detail?.state || STATE_IDLE
        this.syncControls()
      }

      syncControls () {
        if (this.hasButtonTarget) {
          this.buttonTarget.disabled = this.isLoading
        }
      }

      handleActionEvent (event) {
        if (!event) return

        event.preventDefault()
        if (event.target?.blur) event.target.blur()
      }

      get groupElement () {
        return this.element.closest('.f-ai-c-text-suggestions-group')
      }

      get initialGroupState () {
        return this.groupElement?.dataset.fAiCTextSuggestionsGroupStateValue || STATE_IDLE
      }

      get isLoading () {
        return this.groupState === STATE_LOADING
      }
    })
  }

  if (window.Folio?.Stimulus?.register && window.Stimulus?.Controller) {
    registerControlsController()
  } else {
    document.addEventListener('folio:stimulus-ready', registerControlsController, { once: true })
  }
})()
